### PR TITLE
Introduce support for integer datatype

### LINKED ISSUE #317: Support for `integer` datatype removed from loopback\strong-remoting
IBM support sent me this way. I had created IBM service request 46011999000 because this is for GoDaddy and we have a paid subscription.

loopback version 2.18.1
strong-remoting version 2.19.0
 
converting parameter to target type includes explicit support for 'integer' in shared-method.js
lines 390-443

```node 
function convertToBasicRemotingType(type) {
  if (Array.isArray(type)) {
    return type.map(convertToBasicRemotingType);
  }
 
  if (typeof type === 'object') {
    type = type.modelName || type.name;
  }
 
  type = String(type).toLowerCase();
 
  switch (type) {
    case 'string':
    case 'number':
    case 'date':
    case 'boolean':
    case 'buffer':
    case 'object':
    case 'any':
      return type;
    case 'integer':
      return 'number';
    case 'array':
      return ['any'].map(convertToBasicRemotingType);
    default:
      // custom types like MyModel
      return 'object';
  }
}
 
function convertValueToTargetType(argName, value, targetType) {
  switch (targetType) {
    case 'string':
      return String(value).valueOf();
    case 'date':
      return new Date(value);
    case 'number':
    case 'integer':
      return Number(value).valueOf();
    case 'boolean':
      return Boolean(value).valueOf();
    // Other types such as 'object', 'array',
    // ModelClass, ['string'], or [ModelClass]
    default:
      switch (typeof value) {
        case 'string':
          return JSON.parse(value);
        case 'object':
          return value;
        default:
          throw new badArgumentError(argName + ' must be ' + targetType);
      }
  }
}
```
 
loopback version 2.28.0
strong-remoting version 2.26.0
 
converting parameter to target type does not include support for 'integer' in shared-method.js
lines 373-424
 
```node
function convertToBasicRemotingType(type) {
  if (Array.isArray(type)) {
    return type.map(convertToBasicRemotingType);
  }
 
  if (typeof type === 'object') {
    type = type.modelName || type.name;
  }
 
  type = String(type).toLowerCase();
 
  switch (type) {
    case 'string':
    case 'number':
    case 'date':
    case 'boolean':
    case 'buffer':
    case 'object':
    case 'file':
    case 'any':
      return type;
    case 'array':
      return ['any'].map(convertToBasicRemotingType);
    default:
      // custom types like MyModel
      return 'object';
  }
}
 
function convertValueToTargetType(argName, value, targetType) {
  switch (targetType) {
    case 'string':
      return String(value).valueOf();
    case 'date':
      return new Date(value);
    case 'number':
      return Number(value).valueOf();
    case 'boolean':
      return Boolean(value).valueOf();
    // Other types such as 'object', 'array',
    // ModelClass, ['string'], or [ModelClass]
    default:
      switch (typeof value) {
        case 'string':
          return JSON.parse(value);
        case 'object':
          return value;
        default:
          throw new badArgumentError(argName + ' must be ' + targetType);
      }
  }
}
```

We have Swagger specs that have been using integer which is completely valid but a newer version of loopback we'd be forced to switch to number and manually handle checking for fractional values. I don't think that's appropriate.

### PR DESCRIPTION
**Pull request by gunjpan**
all types of numbers, in Javascript, are of only one dataType: `Numbers`.
~~This is a POC~~ Patch to fix #317.
~~Note: Breaks few test cases on local~~

Connect to #317

### CODE REVIEW

**Comment by gunjpan** [comment]
@bajtos : Here is the PoC for introducing `integer` data types. PTAL and let's discuss if the direction is correct. Please note, CI will fail as it had six failing test cases on local, which will be fixed if we're moving further this direction

File: lib/shared-method.js [hunk 1]
```diff
@@ -440,16 +444,17 @@ function convertValueToTargetType(argName, value, targetType) {
...
 
   switch (type) {
     case 'undefined':
     case 'boolean':
-    case 'number':
     case 'function':
     case 'string':
       return type;
+    case 'number':
+      return Number.isInteger(val) ? (targetType || 'integer') : 'number';
```
Comment (gunjpan) [inline lib/shared-method.js]: that's the hacky way to work with numbers which are of pattern: `x.0`, for example: `2.0`

File: lib/shared-method.js [hunk 2]
```diff
@@ -369,7 +369,9 @@ function coerceAccepts(uarg, desc) {
   if (actualType === 'number' && Number.isNaN(uarg)) {
     throw new badArgumentError(name + ' must be a number');
   }
-
+  if (actualType === 'integer' && Number.isNaN(uarg)) {
```
Comment (bajtos) [inline lib/shared-method.js]: I think we need to check `Number.isInteger` here, shouldn't we?

File: lib/shared-method.js (hunk 1 above)
Comment (bajtos) [inline lib/shared-method.js]: Can we change/simplify this condition please? I find it difficult to understand correctly.

```js
return targetType === 'integer' && Number.isInteger(val) ? 'integer' : 'number';
```

Also please add a unit-test to describe how `number` value `2.0` is handled.

File: test/rest.test.js [hunk 3]
```diff
@@ -2253,6 +2276,8 @@ describe('strong-remoting-rest', function() {
   });
 
   it('rejects multi-item array passed to a number argument', function(done) {
+    /* it rejects multi-item array passed to integer argument
+     as it coerces to number type */
```
Comment (bajtos) [inline test/rest.test.js]: Is this comment pointing out a missing unit-test?

File: test/shared-method.test.js [hunk 4]
```diff
@@ -146,6 +146,50 @@ describe('SharedMethod', function() {
       });
     });
 
+    it('returns 400 when integer argument is float number',
+      function(done) {
+        var method = givenSharedMethod({
+          accepts: [{ arg: 'num', type: 'integer' }],
+        });
+
+        method.invoke('ctx', { num: NaN }, function(err) {
```
Comment (bajtos) [inline test/shared-method.test.js]: `NaN` is not a flat number, please use a real float value here, e.g. `1.2`.

File: lib/shared-method.js (hunk 1 above)
Comment (bajtos) [inline lib/shared-method.js]: > Also please add a unit-test to describe how number value 2.0 is handled.

Ah, is this covered by "treats number argument of type x.0 as number"?

**Comment by bajtos** [comment]
@gunjpan Looks mostly good 👍 Please add a unit-test for "array of integers", address my comments above and fix test failures. I think adding support for "integer" should not be a breaking change, i.e. all existing tests should keep passing.

File: test/rest.test.js (hunk 3 above)
Comment (gunjpan) [inline test/rest.test.js]: yes, IIRC. As it will be an exact similar code block, will include the test case for `integer`

File: test/shared-method.test.js (hunk 4 above)
Comment (gunjpan) [inline test/shared-method.test.js]: My bad, I forgot to change after copying from the other case.

File: lib/shared-method.js (hunk 1 above)
Comment (gunjpan) [inline lib/shared-method.js]: Yes, that's correct.

File: lib/shared-method.js (hunk 2 above)
Comment (gunjpan) [inline lib/shared-method.js]: No, as there is already a check for the data type at [line#352](https://github.com/strongloop/strong-remoting/pull/323/files#diff-157ee1a11c353da754b459ce1fe0bc0dR352). This is for `NaN` values only.

File: lib/shared-method.js [hunk 5]
```diff
@@ -576,12 +592,36 @@ SharedMethod.toResult = function(returns, raw, ctx) {
...
+//     var result = x;
+//     if (targetType === 'integer') {
+//       result = convertToIntegerTargetType(x);
+//       this.update(result);
+//     }
+//     return result;
+//   });
+//   return data;
+// }
+
```
Comment (gunjpan) [inline lib/shared-method.js]: Experimental

File: test/rest.test.js [hunk 6]
```diff
@@ -950,29 +950,85 @@ describe('strong-remoting-rest', function() {
...
+            {
+              accepts: { arg: 'arg', type: ['number'] },
+              returns: { arg: 'data', type: ['integer'] },
+              http: { method: 'POST' },
+            });
+
+          request(app).post(method.url)
+            .send({ arg: [1.5, 2] })
+            .expect(200, { data: [3] })
+            .end(done);
```
Comment (gunjpan) [inline test/rest.test.js]: This test-case is failing ATM, thus skipped.

**Comment by bajtos** [comment]
@gunjpan there is one more thing we will need to do as part of fixing the issue, and that's adding new tests to rest-coercion integration tests in `test/rest-coercion`. I slightly prefer to do it as part of this patch, to ensure we don't miss any edge case here, but I won't mind if you decide to do that in a follow-up pull request.

File: lib/shared-method.js (hunk 2 above)
Comment (bajtos) [inline lib/shared-method.js]: > No, as there is already a check for the data type at line#352.

Is my understanding correct that for non-integer numeric value, `actualType` will be `number`, and thus we will reject the value on the grounds that `number` is not `integer`? That makes sense.

Please add a unit-test to cover this case.

File: lib/shared-method.js [hunk 7]
```diff
@@ -551,21 +558,34 @@ SharedMethod.toResult = function(returns, raw, ctx) {
...
-  function convert(val) {
-    switch (SharedMethod.getType(val)) {
+  function convert(val, targetType) {
+    var targetTypeIsArray = Array.isArray(targetType) &&
+      targetType.length === 1;
+
+    if (targetTypeIsArray && Array.isArray(val)) {
+      var arrayTargetType = targetType[0];
+      return val.map(function(arg, ix) {
+        return convert(arg, arrayTargetType);
```
Comment (bajtos) [inline lib/shared-method.js]: I think this is a breaking change, because we were not converting date/buffer types in an array before. I.e. array items were serialized using default `JSON.serialize` convention before, this convention is different from what strong-remoting implements. See https://github.com/strongloop/loopback/issues/1990 and https://github.com/strongloop/loopback/issues/1907 for more context.

I agree the new behaviour is better, because it's what most people would expect, but I am not sure if we can back-port such change to 2.x.

I am proposing to rework `convert` method to ignore all arrays except "array of integers", in order to preserve backwards compatibility, and then address the issue of encoding of array items later.

@gunjpan @ritch thoughts?

Comment (bajtos) [inline lib/shared-method.js]: Clarification of what I meant by my comment:

```
integer - do convert
date - do convert
buffer - do convert
array of integers - do convert
array of dates - no conversion
array of buffers - no conversion
```

Comment (gunjpan) [inline lib/shared-method.js]: @bajtos: Yeah, I agree with 
> rework convert method to ignore all arrays except "array of integers"

File: lib/shared-method.js [hunk 8]
```diff
@@ -566,20 +572,20 @@ SharedMethod.toResult = function(returns, raw, ctx) {
...
+    var targetTypeIsIntegerArray = Array.isArray(targetType) &&
+      targetType.length === 1 && targetType[0] === 'integer';
 
-    if (targetTypeIsArray && Array.isArray(val)) {
+    if (targetTypeIsIntegerArray && Array.isArray(val)) {
       var arrayTargetType = targetType[0];
-      return val.map(function(arg, ix) {
-        return convert(arg, arrayTargetType);
+      return val.map(function(intVal) {
+        return convert(intVal, arrayTargetType);
```
Comment (bajtos) [inline lib/shared-method.js]: Should you forward `name` arg to `convert()` call here?

Comment (bajtos) [inline lib/shared-method.js]: Sorry, I meant `argName`.

File: lib/shared-method.js [hunk 9]
```diff
@@ -596,14 +602,19 @@ SharedMethod.toResult = function(returns, raw, ctx) {
...
-      case 'number':
+      default:
         if (targetType === 'integer') {
-          if (!Number.isInteger(val)) return parseInt(val, 10);
+          if (!Number.isInteger(val)) {
+            var message = util.format(
+              'Invalid return value for argument \'%s\' of type ' +
+              '\'%s\': %s. Received type was %s.',
+              argName, targetType, val, typeof val);
+            throw new internalServerError(message);
```
Comment (bajtos) [inline lib/shared-method.js]: `internalServerError` already calls `new`, no need to call it again here.

File: test/rest-coercion/jsonform-integer.suite.js [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+      // Empty values
+      [EMPTY_BODY, undefined],
+      [{ arg: null }, 0], // should be: null
+
+      // Valid values
+      [{ arg: 0 }, 0],
+      [{ arg: 1 }, 1],
+      [{ arg: -1 }, -1],
+
+      // Integers larger than MAX_SAFE_INTEGER get trimmed
```
Comment (bajtos) [inline test/rest-coercion/jsonform-integer.suite.js]: IMO, integers larger than MAX_SAFE_INTEGER should be rejected with 400 Bad Request. We can do that change outside of this pull request if you prefer, but the comment should be changed in such case.

File: test/rest-coercion/jsonform-integer.suite.js [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+      [{ arg: 2343546576878989879789 }, 2.34354657687899e+21],
+      [{ arg: -2343546576878989879789 }, -2.34354657687899e+21],
+
+      // Scientific notation works
+      [{ arg: 1.234e+30 }, 1.234e+30],
+      [{ arg: -1.234e+30 }, -1.234e+30],
+
+      // Integer-like string values should not trigger ERROR_BAD_REQUEST
```
Comment (bajtos) [inline test/rest-coercion/jsonform-integer.suite.js]: Ideally, I would expect that JSON request body `{ arg: '0' }` triggers 400 Bad request when `arg` is expected to be a number. The test suite for numbers ([link](https://github.com/strongloop/strong-remoting/blob/27db434b7b252efbfea7128142e5b49a49eb16a5/test/rest-coercion/jsonform-number.suite.js#L53)) says "Number-like string values should trigger ERROR_BAD_REQUEST", I don't see why integers should behave differently.

File: test/rest-coercion/jsonform-integer.suite.js [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+      [{ arg: '0' }, 0],
+      [{ arg: '1' }, 1],
+      [{ arg: '-1' }, -1],
+      [{ arg: '2343546576878989879789' }, 2.34354657687899e+21],
+      [{ arg: '-2343546576878989879789' }, -2.34354657687899e+21],
+      [{ arg: '1.234e+30' }, 1.234e+30],
+      [{ arg: '-1.234e+30' }, -1.234e+30],
+
+      // All other non-integer values should trigger ERROR_BAD_REQUEST
+      [{ arg: '1.2' }, ERROR_BAD_REQUEST],
```
Comment (bajtos) [inline test/rest-coercion/jsonform-integer.suite.js]: You are missing few test cases here for values that should be rejected:

```js
{ arg: 1.2 },
{ arg -1.2 },
```

File: test/rest-coercion/urlencoded-integer.suite.js [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+      [EMPTY_QUERY, undefined],
+      ['arg', 0], // should be: undefined
+      ['arg=', 0], // should be: undefined
+
+      // Valid values
+      ['arg=0', 0],
+      ['arg=1', 1],
+      ['arg=-1', -1],
+
+      // Numbers larger than MAX_SAFE_INTEGER get trimmed
```
Comment (bajtos) [inline test/rest-coercion/urlencoded-integer.suite.js]: Should be rejected, see https://github.com/strongloop/strong-remoting/pull/323/files/e8649ef1693273e0c8fa698fa7c7bfc70d422552..e066d0e76c608e014d312f6efac0281c08392bc0#r72065917

File: test/shared-method.test.js [hunk 14]
```diff
@@ -202,8 +203,29 @@ describe('SharedMethod', function() {
...
-    });
 
+      it('returns 500 for non-integer return value if type: `integer`',
+        function(done) {
+          var method = givenSharedMethod(
+            function(cb) {
+              cb(null, 3.141);
+            },
+            {
+              returns: [
```
Comment (bajtos) [inline test/shared-method.test.js]: Nitpick: L214-L216 can be abbreviated to

```js
returns: { arg: 'value', type: 'integer' }
```

File: lib/shared-method.js (hunk 8 above)
Comment (gunjpan) [inline lib/shared-method.js]: yes, it should.

**Comment by gunjpan** [comment]
@bajtos : Previous comments addressed. PTAL at the latest commit. Thanks.

File: lib/shared-method.js [hunk 15]
```diff
@@ -701,3 +716,15 @@ SharedMethod.prototype.addAlias = function(alias) {
...
+
+// Include support for following Number methods to support Node v.0.10.x
+Number.isInteger = Number.isInteger || function(value) {
+  return typeof value === 'number' &&
+    isFinite(value) &&
+    Math.floor(value) === value;
+};
+
+Number.isSafeInteger = Number.isSafeInteger || function(value) {
+  Number.MAX_SAFE_INTEGER = Number.MAX_SAFE_INTEGER || Math.pow(2, 53) - 1;
```
Comment (bajtos) [inline lib/shared-method.js]: Please move this line outside of `Number.isSafeInteger` function, no need to run it every time `isSafeInteger` is called.

File: test/rest-coercion/jsonform-integer.suite.js [hunk 16]
```diff
@@ -40,24 +40,26 @@ module.exports = function(ctx) {
...
       [{ arg: '1' }, 1],
       [{ arg: '-1' }, -1],
-      [{ arg: '2343546576878989879789' }, 2.34354657687899e+21],
-      [{ arg: '-2343546576878989879789' }, -2.34354657687899e+21],
-      [{ arg: '1.234e+30' }, 1.234e+30],
-      [{ arg: '-1.234e+30' }, -1.234e+30],
+      [{ arg: '2343546576878989879789' }, ERROR_BAD_REQUEST],
+      [{ arg: '-2343546576878989879789' }, ERROR_BAD_REQUEST],
+      [{ arg: '1.234e+30' }, ERROR_BAD_REQUEST],
+      [{ arg: '-1.234e+30' }, ERROR_BAD_REQUEST],
```
Comment (bajtos) [inline test/rest-coercion/jsonform-integer.suite.js]: Please add two more items: `'1.234e+3` and `'-1.234e+3'` (strings containing an integer using scientific notation)

**Comment by bajtos** [comment]
Two more comments, you are almost there :)

**Comment by bajtos** [comment]
LGTM. Please squash the commits and wait for CI build to turn green before landing.

**Comment by gunjpan** [comment]
@slnode test please

**Comment by gunjpan** [comment]
@slnode test please

**Comment by gunjpan** [comment]
@bajtos: Landed. Do we need to backport it to 2.x?
